src/nbl/ext/DepthPyramidGenerator/DepthPyramidGenerator.cpp

```diff
@@ -0,0 +1,244 @@
+		writes[0].arrayElement = 0u;
+		writes[0].count = 1u;
+		writes[0].descriptorType = EDT_COMBINED_IMAGE_SAMPLER;
+		writes[0].info = &infos[0];
+
+		writes[1].dstSet = m_ds.get();
+		writes[1].binding = 1;
+		writes[1].arrayElement = 0u;
+		writes[1].count = 1u;
+		writes[1].descriptorType = EDT_STORAGE_IMAGE;
+		writes[1].info = &infos[1];
+
+		m_driver->updateDescriptorSets(sizeof(writes) / sizeof(IGPUDescriptorSet::SWriteDescriptorSet), writes, 0u, nullptr);
+	}
+
+	// create extra shader code
+	constexpr uint32_t maxExtraCodeSize = 256u;
+	std::string defines;
+	defines.reserve(maxExtraCodeSize);
+	{
+		switch (config.workGroupSize)
+		{
+		case E_WORK_GROUP_SIZE::E32x32x1:
+			defines += std::string("#define WORKGROUP_X_AND_Y_SIZE 32\n");
+			break;
+		case E_WORK_GROUP_SIZE::E16x16x1:
+			defines += std::string("#define WORKGROUP_X_AND_Y_SIZE 16\n");
+			break;
+		}
+
+		// TODO: complete supported formats
+		switch (config.outputFormat)
+		{
+		case EF_R16_SFLOAT:
+			defines += std::string("#define MIP_IMAGE_FORMAT r16f\n");
+			break;
+		case EF_R32_SFLOAT:
+			defines += std::string("#define MIP_IMAGE_FORMAT r32f\n");
+			break;
+		case EF_R16G16_SFLOAT:
+			defines += std::string("#define MIP_IMAGE_FORMAT r16g16f\n");
+			break;
+		case EF_R32G32_SFLOAT:
+			defines += std::string("#define MIP_IMAGE_FORMAT r32g32f\n");
+			break;
+		default:
+			assert(false);
+		}
+
+		if (!config.roundUpToPoTWithPadding)
+			defines += std::string("#define STRETCH_MIN\n");
+	}
+
+	asset::IAssetLoader::SAssetLoadParams lp;
+	auto shader = IAsset::castDown<ICPUSpecializedShader>(*am->getAsset("../../../include/nbl/builtin/glsl/ext/DepthPyramidGenerator/depth_pyramid_generator.comp", lp).getContents().begin());
+	assert(shader);
+	const asset::ICPUShader* unspec = shader->getUnspecialized();
+	assert(unspec->containsGLSL());
+
+	//	override shader code
+	auto begin = reinterpret_cast<const char*>(unspec->getSPVorGLSL()->getPointer());
+	const std::string_view origSource(begin, unspec->getSPVorGLSL()->getSize());
+
+	const size_t firstNewlineAfterVersion = origSource.find("\n", origSource.find("#version "));
+	assert(firstNewlineAfterVersion != std::string_view::npos);
+	const std::string_view sourceWithoutVersion(begin + firstNewlineAfterVersion, origSource.size() - firstNewlineAfterVersion);
+
+	std::string newSource("#version 460 core\n");
+	newSource += defines;
+	newSource += sourceWithoutVersion;
+
+	auto unspecNew = core::make_smart_refctd_ptr<asset::ICPUShader>(newSource.c_str());
+	auto specinfo = shader->getSpecializationInfo();
+
+	auto newSpecShader = core::make_smart_refctd_ptr<asset::ICPUSpecializedShader>(std::move(unspecNew), std::move(specinfo));
+	auto gpuShader = driver->getGPUObjectsFromAssets(&newSpecShader.get(), &newSpecShader.get() + 1u)->begin()[0];
+
+	m_ppln = m_driver->createGPUComputePipeline(nullptr, m_driver->createGPUPipelineLayout(nullptr, nullptr, core::smart_refctd_ptr(dsLayout)), std::move(gpuShader));
```

Comment:
this is nasty, other extensions do it better (create a temporary GLSL file, that then `#includes` the builtins properly).